walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,79 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<String>(request.getAttendees());
+    List<TimeRange> initalRanges = new ArrayList<TimeRange>();
+    List<TimeRange> results = new ArrayList<TimeRange>();
+    final int minMeetingTime = 15;
+    
+    for(int start = 0; start <= TimeRange.END_OF_DAY; start += minMeetingTime) {
+        if(start + minMeetingTime > TimeRange.END_OF_DAY) {
+            initalRanges.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, false));
+        } else {
+            initalRanges.add(TimeRange.fromStartDuration(start, minMeetingTime));
+        }
+    }
+
+    for(Event currEvent : events) {
+        Set<String> currAttendees = currEvent.getAttendees();
+        for(String attendee : currAttendees) {
+            if(attendees.contains(attendee)) {
+                blockTimeRange(initalRanges, currEvent);
+                break;
+            }
+        }
+    }
+
+    TimeRange tempRange = new TimeRange();
+    int duration = (int) request.getDuration();
+
+// Group together the leftover time ranges for results
+    for(TimeRange range : initalRanges) {
+
+        //range is not schedulable and there was a previous valid range started
+        if(!range.isSchedulable() && tempRange.isValidRange()) {
+            if(tempRange.getDuration() >= duration) {
+            results.add(TimeRange.fromStartDuration(tempRange.getStart(), tempRange.getDuration()));
+            }
+            tempRange.setValidRange(false);
+            tempRange.setStart(0);
+            tempRange.setDuration(0);
+        }
+
+        //range is schedulable and there is a valid range in the works
+        if(range.isSchedulable() && tempRange.isValidRange()) {
+            tempRange.setDuration(tempRange.getDuration() + range.getDuration());
+        }
+
+        //range is schedulable and there is not a vaild range already in the works then start a valid range
+        if(range.isSchedulable() && !tempRange.isValidRange()) {
+            tempRange.setValidRange(true);
+            tempRange.setStart(range.getStart());
+            tempRange.setDuration(range.getDuration());
+        }
+
+        //If it is the last range of the initial ranges and there is a viable range in the works and
+        //the range duration statisfies the required duration then add it to results
+        if(range.end() == TimeRange.END_OF_DAY && tempRange.isValidRange() && tempRange.getDuration() >= duration) {
```

Comment:
Instead of doing this here, do this after the loop. That way it's clearer and does not need the extra if.

ie,
for (TimeRange range : ranges) {
  // your code
}
// Handle the last time range
maybeAddToResult(tempRange, resultList)